pennylane/gradients/general_shift_rules.py

```diff
@@ -0,0 +1,131 @@
+
+    Returns:
+         tuple: a tuple of one nested list describing the gradient recipe
+            for the parameter-shift method.
+            This is a tuple with one nested list per operation parameter. For
+            parameter :math:`\phi_k`, the nested list contains elements of the form
+            :math:`[c_i, a_i, s_i]` where :math:`i` is the index of the
+            term, resulting in a gradient recipe of
+
+            .. math:: \frac{\partial}{\partial\phi_k}f = \sum_{i} c_i f(a_i \phi_k + s_i).
+
+
+    Raises:
+        ValueError: if `frequencies` is not a list of unique positive values, or if `shifts`
+            (if specified) is not a list of unique values the same length as `frequencies`.
+
+    **Examples**
+
+    An example of obtaining the frequencies from a set of unique eigenvalues and obtaining the
+    parameter shift rule:
+
+    >>> unique_eigenvals = [1, -1, 0]
+    >>> unique_eigenvals = sorted(unique_eigenvals)
+    >>> frequencies = set()
+    >>> for i in range(len(unique_eigenvals)):
+    >>>     for j in range(i+1, len(unique_eigenvals)):
+    >>>         frequencies.add(unique_eigenvals[j] - unique_eigenvals[i])
+    >>> get_shift_rule(tuple(frequencies))
+    ([[0.8535533905932737, 1, 0.7853981633974483], [-0.14644660940672624, 1, 2.356194490192345],
+    [-0.8535533905932737, 1, -0.7853981633974483], [0.14644660940672624, 1, -2.356194490192345]],)
+
+    An example with explicitly specified shift values:
+
+    >>> frequencies = (1, 2, 4)
+    >>> shifts = (np.pi / 3, 2 * np.pi / 3, np.pi / 4)
+    >>> get_shift_rule(frequencies, shifts)
+    ([[-2.0907702751760278, 1, 1.0471975511965976], [0.2186308015824754, 1, 2.0943951023931953],
+    [3.0000000000000004, 1, 0.7853981633974483], [2.0907702751760278, 1, -1.0471975511965976],
+    [-0.2186308015824754, 1, -2.0943951023931953], [-3.0000000000000004, 1, -0.7853981633974483]],)
+    """
+
+    n_freqs = len(frequencies)
+
+    if len(set(frequencies)) != n_freqs or any(freq <= 0 for freq in frequencies):
+        raise ValueError(
+            "Expected frequencies to be a list of unique positive values, instead got {}.".format(
+                frequencies
+            )
+        )
+
+    if shifts is None:  # assume equidistant shifts
+        mu = np.arange(1, n_freqs + 1)
+        shifts = (2 * mu - 1) * np.pi / (2 * n_freqs)
+    else:
+        shifts = qml.math.stack(shifts)
+        if len(shifts) != n_freqs:
+            raise ValueError(
+                "Expected number of shifts to equal the number of frequencies ({}), instead got {}.".format(
+                    n_freqs, shifts
+                )
+            )
+        if len(set(shifts)) != n_freqs:
+            raise ValueError("Shift values must be unique, instead got {}".format(shifts))
+
+    frequencies = qml.math.sort(qml.math.stack(frequencies))
+    freq_min = frequencies[0]
+
+    if len(set(np.diff(frequencies))) == 1:  # equidistant case  # equidistant case

```

Comment:
Hey @ral9000, I just noticed that this gives a very odd result for the 2-term shift rule:

```python
>>> get_shift_rule([2])  # since single qubit rotations have eigvals [1, -1]
[[ 1.57079633e+00 -1.57079633e+00]
 [ 8.16561968e+15 -8.16561968e+15]]
```

I think this is because the diff in the frequencies is an empty list!

So this can be fixed by doing 

```suggestion
    if len(set(np.diff(frequencies))) == (0, 1):  # equidistant case
```

However, I then get (for coeffs, shifts):

```python
>>> get_shift_rule([2])
[[ 0.78539816 -0.78539816]
 [ 1.         -1.        ]]
```

I think you need to remove the `freq_min` division/multiplication in the equidistant shift case? If that is removed, I then get the expected result:


```python
>>> get_shift_rule([2])
[[ 0.5        -0.5       ]
 [ 1.57079633 -1.57079633]]
```

Reply:
Thanks @josh146 for finding that catch with the 1 frequency equidistant case.  However, I'm unsure of the modification. Currently the parameter shift rule output by `get_shift_rule` agrees with Eq.(14) of [arXiv:1811.11184](https://arxiv.org/abs/1811.11184v1),  where the shifts are `\pm pi/(4r)` with coefficients `\pm r`, for generator with spectrum `\pm r`, i.e. has a single frequency `2r`.

>>> get_shift_rule((2,))
([[1.0, 1, 0.7853981633974483], [-1.0, 1, -0.7853981633974483]],)
>>> get_shift_rule((1,))
([[0.5, 1, 1.5707963267948966], [-0.5, 1, -1.5707963267948966]],)

When deriving how the general equidistant shift rules for a frequency spectrum (1,2,...,R) should be modified to handle other equidistant cases, I saw that the coefficients should be reweighted by `freq_min`, and shifts should be reweighted by `1 / freq_min`. It looks like your suggestion is to reweight using `freq_min / 2` and `2 / freq_min` respectively. I believe doing so won't always give the correct result, e.g.:

`import pennylane as qml

n_wires = 1
dev = qml.device("default.qubit", wires=n_wires)

hamiltonian = qml.Hamiltonian([1], [qml.PauliX(0)])
frequencies = (2,)

@qml.qnode(dev)
def circuit(time):
    qml.PauliX(0)
    qml.CommutingEvolution(hamiltonian, time, frequencies)
    return qml.expval(qml.PauliZ(0))

def grad_rule_standard(x):
    grad_recon = 0
    for term in qml.gradients.get_shift_rule(frequencies)[0]:
        grad_recon += term[0]*circuit(x+term[2])
    return grad_recon

def grad_rule_reweight(x):
    grad_recon = 0
    for term in qml.gradients.get_shift_rule(frequencies)[0]:
        grad_recon += 1/2*term[0]*circuit(x+2*term[2])
    return grad_recon

print(qml.gradients.finite_diff(circuit)(0.4))
print(grad_rule_standard(0.4))
print(grad_rule_reweight(0.4))`

gives output:

`1.4347122490043773
1.4347121817990454
1.1102230246251565e-16`

Reply:
@ral9000 ah I think I'm starting to see the reason I was adding in the `1/2` term - I was thinking of rotation gates such as `RX`, `CRX`, etc., where the generator is multiplied by a scalar of `1/2` in the exponent.

In this case, the rescaling is needed. E.g., consider the following Hamiltonian:

```python
>>> hamiltonian = qml.Hamiltonian([0.5], [qml.PauliX(0)])
>>> @qml.qnode(dev)
... def circuit(time):
...     qml.PauliX(0)
...     qml.CommutingEvolution(hamiltonian, time, frequencies)
...     return qml.expval(qml.PauliZ(0))
>>> print(qml.gradients.finite_diff(circuit)(0.4))
[[0.38941839]]
>>> print(grad_rule_standard(0.4))
0.5507207011297415
>>> print(grad_rule_reweight(0.4))
0.3894183423086503
```

So it looks like we need to take the generator coefficient into account when computing the general shift rule?

Reply:
That's right, the generator coefficient is taken into account via `frequencies`. Now eigenvalues of the generator are `pm 1/2` (as opposed to `pm 1`), so `frequencies = (1,)`, which will give the correct gradient at x via `grad_rule_standard`

Reply:
Ah, no, I realized that the frequencies just needs to be reduced to `(1,)` 🤦 Ignore my train of thought above!

Reply:
you beat me too it 😆